Ignore check

gbataille opened the pull request:
No tests (yet) because I'm lazy but it works with the small checks I made.

Can you do a few tests on your massive codebase?

If you agree with the design and the performance, I'll finish adding tests.

### Note

There is space to add `IgnoreCommand` like `krank:ignore-file` on the top of a file or `krank:ignore-xxx` to ignore a specific checker on the tagged line

gbataille committed:
refactor: SourcePos -> record syntax for "documentation"

gbataille committed:
refactor: Localized in Krank.Types

gbataille committed:
refactor: import qualified of Regex

gbataille committed:
refactor: explicit imports

gbataille committed:
refactor: reorder imports

gbataille committed:
feat: ignore line

gbataille committed:
feat: ignore issues on line

guibou commented on src/Krank/Checkers/Ignore.hs:
```diff
@@ -0,0 +1,76 @@
+-- | This regex represents a krank ignore marker
+ignoreRe :: RE.Regex
+-- NOTE: \b at the beginning is really import for performances
+-- because it dramatically reduces the number of backtracks
```
Technically here that's not true. It reduces the number of backtrack in the url parser because the url parser starts by an optional string and then matchs an unspecified number of arbitrary characters.

This means that the regex engine must start by trying the optional string. If it succeed, it will then match any unspecified number of arbitrary characters. If by any bad luck the match fails, it will come back the the beginning and try without the optional string and do the same process. And this is done for any offset in the input string.

By forcing the first char of the match to be something specific (i.e. a word boundary `\b`), the regex engine first start by looking for a word boundary, then it does the complex analysis with possible backtracking. The change here is important, because instead of applying the costly process on all offset of the input string, it only applies it on all offsets which match the specific start pattern (i.e. a word boundary). The ratio of word boundaries over characters in a text is easily 5 (I assume that it may be the mean of the length of words in a standard piece of program.

guibou reviewed (changes requested):
I'll design that differently. Something such as:

```diff
diff --git a/src/Krank/Checkers/IssueTracker.hs b/src/Krank/Checkers/IssueTracker.hs
index 72def02..1a1d717 100644
--- a/src/Krank/Checkers/IssueTracker.hs
+++ b/src/Krank/Checkers/IssueTracker.hs
@@ -72,7 +72,9 @@ gitRepoRe = [re|\b(?>https?://)?(?>www\.)?([^/]+)/([^/]+)/((?>[^/]+))/issues/([0
 
 -- | Extract all issues on one line and returns a list of the raw text associated with an issue
 extractIssuesOnALine :: ByteString -> [(Int, GitIssue)]
-extractIssuesOnALine lineContent = map f (scan gitRepoRe lineContent)
+extractIssuesOnALine lineContent
+  | "krank:ignore" `Text.isInfixOf` lineContent = []
+  | otherwise = map f (scan gitRepoRe lineContent)
       where
         f (match, [domain, owner, repo, ByteString.readInt -> Just (issueNo, _)]) = (colNo, GitIssue provider (Text.Encoding.decodeUtf8 owner) (Text.Encoding.decodeUtf8 repo) issueNo)
           where
```

Which I admit is more coupled to the issue checker algorithm, but on the other hand, it is really simpler.

gbataille committed:
refactor: remove \b at start of the RE

As @guibou explained, here the `\b` does not really improve performance.
It was extremely useful in the case where the RE pattern to match was
starting by an optional string, which is not the case here.

gbataille commented:
I don't agree.
For me the whole point of the tool is to be a linter à la `pylint` for example. So the IssueTracker is just one among many checks that could be ignored.
I agree that your way is much more efficient but that would force each checker to implement its ignore logic which I don't find to be good design.

guibou commented:
I do agree with your point.

We may refactor later in a way which streams all the file line by line and pass each lines to a checker and to the ignore list. Or we may not.

Did you consider the PR is a mergeable state now or do you want to add test beforehand?

gbataille commented:
Yeah, going line by line, with ignore first would be an interesting change, but further preventing (or at least making harder) the handling of blocks. To be seen.

I want to add tests, but I think the ignore + the gitlab you have done were the 2 main issues people had with the lib. 

Knowing it seems to work (manual tests), it's simple, and done. I would say we merge, do a proper release (maybe with an announcement on a few channel) and then we start caring about quality and additional features.

Do you agree?

guibou commented:
I do manual test today and I merge. THANK YOU!

guibou commented:
No impact on timing on my huge codebase. Other than that, it works! Thank you.
